### PR TITLE
Open PR in Simorgh when Talos is run successfully

### LINKED ISSUE #2174: Allow Talos to raise a PR in Simorgh updating `@bbc/psammead*` pacakges
**Is your feature request related to a problem? Please describe.**
Now that we have talos successfully keeping `@bbc/psammead*` packages up-to-date we should enable talos to open a PR in Simorgh so it can update the newly published psammead packages.

**Describe the solution you'd like**
Once talos has merged a PR updating psammead dependencies it should also raise a PR in simorgh to update to the newly published dependencies.

Thus creating 3 PR when making changes in psammead.

1) Initial psammead change
2) Auto talos PR updating bbc deps
3) (New) Auto talos pr updating `@BBC/psammead*` deps in Simorgh

**Describe alternatives you've considered**
Not doing this..............

**Testing notes**
[Tester to complete]

Dev insight: Will Cypress tests be required or are unit tests sufficient? Will there be any potential regression? etc

- [ ] This feature is expected to need manual testing.

**Additional context**

### PR DESCRIPTION
**Pull request by pharingee**
Resolves #2174

**Overall change:** 
Remotely create a PR in Simorgh when Talos is run successfully to update published packages

**Code changes:**
- Remotely update package.json in Simorgh

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] Automated jest tests added (for new features) or updated (for existing features)
- [ ] This PR requires manual testing

### CODE REVIEW

File: .gitignore
Comment (jroebu14) [inline .gitignore:23]: was this intended?

File: scripts/talos/bumpSimorghPackages/getPublishedPackages/index.js
Comment (jroebu14) [inline scripts/talos/bumpSimorghPackages/getPublishedPackages/index.js]: does this mean the package was upgraded? If so then `isUpgraded` would be a good variable name to put here. It also means the `bumpedPackage` isn't an accurate name

File: .gitignore
Comment (pharingee) [reply .gitignore:23]: Yes it was. There are no vscode files in the codebase and I needed a couple of settings set up which might not be the same for everyone

File: scripts/talos/bumpSimorghPackages/getPublishedPackages/index.js
Comment (pharingee) [reply scripts/talos/bumpSimorghPackages/getPublishedPackages/index.js]: Yeah, I think `publishedPackage` will be a better name

Comment (jroebu14) [reply scripts/talos/bumpSimorghPackages/getPublishedPackages/index.js]: Not a big deal but you can easily extract the things you need with regexes

```js
packageList.forEach(package => {
  const isUpgraded = package.inclues('  →  ');

  if (isUpgraded) {
    const [packageName] = package.match(/^@bbc\/psammead-\S+/)
    const [version] = package.match(/(?<=→\s\s).+$/)
    publishedPackages[packageName] = version;
  }
});
```

File: scripts/talos/bumpSimorghPackages/getUpdates/index.js
Comment (jroebu14) [inline scripts/talos/bumpSimorghPackages/getUpdates/index.js]: No big deal but I try to discourage the use of mutation where possible. I think `map` and `filter(Boolean)` would be cleaner than mutating a variable outside the scope of a `forEach` callback

Comment (pharingee) [reply scripts/talos/bumpSimorghPackages/getUpdates/index.js]: Yeah, initially updates needed to be an object thus the mutation. I'll refactor to array functions now.

File: scripts/talos/bumpSimorghPackages/getPublishedPackages/index.js
Comment (pharingee) [reply scripts/talos/bumpSimorghPackages/getPublishedPackages/index.js]: I tend to find regexes not very straightforward to read. So I avoid them when I can. In this case it seems to make the code more readable so I'll use them.

Comment (jroebu14) [reply scripts/talos/bumpSimorghPackages/getPublishedPackages/index.js]: @pharingee that's a fair point!

File: scripts/talos/bumpSimorghPackages/updatePackageJson/index.js
Comment (jroebu14) [inline scripts/talos/bumpSimorghPackages/updatePackageJson/index.js]: it would be nice to see this stored as a well named variable like `hasNewDependency`

File: scripts/talos/bumpSimorghPackages/updatePackageJson/index.js
Comment (jroebu14) [inline scripts/talos/bumpSimorghPackages/updatePackageJson/index.js:14]: I think the `mergeUpdates` abstraction should be the callback function in the forEach here. It's also possible to build up a single object using a reduce function instead of creating an object to mutate. An example of this could be:

```js
const mergeUpdates = publishedPackages => (packageJson, key) => {
  const { dependencies, devDependencies } = packageJson;
  const newDep = dependencies && dependencies[key];
  const newDevDep = devDependencies && devDependencies[key];

  return mergeAll([
    packageJson,
    newDep && {
      dependencies: {
        ...dependencies,
        [key]: publishedPackages[key],
      },
    },
    newDevDep && {
      devDependencies: {
        ...devDependencies,
        [key]: publishedPackages[key],
      },
    },
  ]);
};
```

and used in `updatePackageJson` like:

```js
  const newPackageJson = Object.keys(publishedPackages).reduce(
    mergeUpdates(publishedPackages),
    packageJson,
  );
```

more lines required to the same thing but possibly more readable and doesn't include mutations.

**Review by jroebu14 — COMMENTED** [review]
I haven't tested this out yet but overall the code seems good to me. Nice work. I'll approve after I test it out manually after lunch!

**Review by jroebu14 — APPROVED** [review]

File: scripts/talos/bumpSimorghPackages/updatePackageJson/index.js
Comment (pharingee) [reply scripts/talos/bumpSimorghPackages/updatePackageJson/index.js:14]: I agree on the avoidance on mutation. But I think this sacrifices some readability to achieve that. I feel your code is more elegant but significantly less readable because of multiple complex uses of object destruction and the reduce function. By renaming some of the variables I think the updates I made should make it easier to read.

**Review by thekp — COMMENTED** [review]
I think it would be worth to update the documentation here: https://github.com/bbc/psammead/blob/latest/scripts/talos/README.md
just to briefly note that Talos also opens a PR in Simorgh now

**Review by FK78 — APPROVED** [review]
LGTM!

**Review by thekp — APPROVED** [review]
Thanks for updating the readme.

Do you have an example of Simorgh Talos PR?

File: scripts/talos/bumpSimorghPackages/updatePackageJson/index.js
Comment (dr3) [inline scripts/talos/bumpSimorghPackages/updatePackageJson/index.js:36]: Why arnt you using non-check-packages for this like psammead?

**Review by dr3 — CHANGES_REQUESTED** [review]
This PR introduces a second way of updating dependencies in a package.json file. IMO we should be consistent and share the logic between updating psammead and simorgh. On my phone ATM, will be able to give more detailed feedback after 4pm BST

**Comment by pharingee** [comment]
@dr3 I agree. I think we should update Talos' run on psammead to drop the use of `npm-check-updates` since it's resource intensive and takes a really long time to complete (In the order of minutes). A sample run of the new method can be seen here on a fork: https://github.com/pharingee/simorgh/pull/4

**Comment by dr3** [comment]
> @dr3 I agree. I think we should update Talos' run on psammead to drop the use of `npm-check-updates` since it's resource intensive and takes a really long time to complete (In the order of minutes). A sample run of the new method can be seen here on a fork: [pharingee/simorgh#4](https://github.com/pharingee/simorgh/pull/4)

@pharingee The fact its slow or resource intensive isnt a concern to me, as talos isnt something whose duration affects developers. I would much rather we use the script, in favour of writing (and having to maintain) our own code.

**Comment by pharingee** [comment]
I think the speed and resources used are important because of 3 reasons:
1. There has been at least one recorded instance where Talos has been aborted by a member of the team.
2. A request was made I think by @amywalker1987 to not fail the Jenkins Pipeline when Talos fails.
3. There are multiple instances of Talos failing at the `npm-check-updates` level.

In respect to maintaining our own code, the amount of code here to maintain is 2 files 51 lines long in total. If we wrote Talos - at close to 1000 lines of maintenance - to fix dependabot, I think the  benefit is worth the amount of code we need to maintain.

File: scripts/talos/bumpSimorghPackages/getUpdates/index.js
Comment (dr3) [inline scripts/talos/bumpSimorghPackages/getUpdates/index.js:16]: Is this being done to mimic the output of `npm-check-updates`? why?

File: scripts/talos/bumpSimorghPackages/index.js
Comment (dr3) [inline scripts/talos/bumpSimorghPackages/index.js:16]: Why have we got 2 ways of creating pull request bodies? (psammead and simorgh prs) when they can be the same?

**Review by dr3 — COMMENTED** [review]
This appears to share almost no commonalities with how we bump psammead packages. Having two entirely different implementations isn't sustainable IMO, 900 lines of code (inc tests to maintain) to do pretty much the same thing, in a new repo seems alot

**Comment by dr3** [comment]
> I think the speed and resources used are important because of 3 reasons:
> 
> 1. There has been at least one recorded instance where Talos has been aborted by a member of the team.

They could equally do that if it was shorter, they shouldnt

> 2. A request was made I think by @amywalker1987 to not fail the Jenkins Pipeline when Talos fails.

Thats not nothing to do with the contents of the talos script, but the way its run.

> 3. There are multiple instances of Talos failing at the `npm-check-updates` level.

Theres also risk in having our own code, if we have issues with the script, our first reaction should be to debug and report bugs to the maintainers, not (half) run away from it

**Comment by pharingee** [comment]
I have open a draft PR that replaces npm-check-updates and completes the check for all packages in 2 minutes here: https://github.com/bbc/psammead/pull/2409

File: scripts/talos/bumpSimorghPackages/index.js
Comment (pharingee) [reply scripts/talos/bumpSimorghPackages/index.js:16]: The two repositories are different in structure. Psammead contains multiple packages, while Simorgh is a single repository. So we need to keep the bodies separate.

File: scripts/talos/bumpSimorghPackages/getUpdates/index.js
Comment (pharingee) [reply scripts/talos/bumpSimorghPackages/getUpdates/index.js:16]: Yes it is. Because there is a function written to take advantage of that structure. If we effectively take `npm-check-updates out`, we can do without the mimicking and have a unified structure across.

Comment (pharingee) [reply scripts/talos/bumpSimorghPackages/getUpdates/index.js:16]: Talos was written to update Psammead. With no intention of updating another repository. The current features of Talos like running commit messages and pushes from the command line will mean that we will have to clone the entire Simorgh repository and isolate it from the current git repository. Which seems like a lot of work which is prone to errors and inconsistency.

Comment (pharingee) [reply scripts/talos/bumpSimorghPackages/getUpdates/index.js:16]: A lot of the new functions written can replace the current way Talos runs by using Github API to make commits and pushes. That will unify both methods.

**Comment by pharingee** [comment]
> They could equally do that if it was shorter, they shouldnt
> 
They do it because it hangs. If it were to throw an error, we could at least debug it. But by hanging for several minutes it's impossible to know what went wrong.


> Theres also risk in having our own code, if we have issues with the script, our first reaction should be to debug and report bugs to the maintainers, not (half) run away from it

`npm-check-updates` was designed to handle published articles to the npm for a package and has been optimised for that. In the case of psammead, having multiple packages updated isn't an ideal use-case for npm-check-updates, which is why we need to couple it with lerna to make it work. This isn't running away but taking a different route which is optimised for our workflow.

**Comment by pharingee** [comment]
Picking this up again

**Pull request closed without merging**
